plugins/tc/tbf/main.go

```diff
@@ -0,0 +1,235 @@
+	conf, err := parseConfig(args.StdinData)
+	if err != nil {
+		return err
+	}
+
+	if conf.Rate == 0 && conf.Burst == 0 {
+		return types.PrintResult(conf.PrevResult, conf.CNIVersion)
+	}
+
+	if conf.PrevResult == nil {
+		return fmt.Errorf("must be called as chained plugin")
+	}
+
+	containerIPConfigs := make([]*current.IPConfig, 0, len(conf.PrevResult.IPs))
+
+	for _, ip := range conf.PrevResult.IPs {
+		if ip.Interface == nil {
+			continue
+		}
+		intIdx := *ip.Interface
+		// Every IP is indexed in to the interfaces array, with "-1" standing
+		// for an unknown interface (which we'll assume to be Container-side
+		// Skip all IPs we know belong to an interface with the wrong name.
+		if intIdx >= 0 && intIdx < len(conf.PrevResult.Interfaces) && conf.PrevResult.Interfaces[intIdx].Name != args.IfName {
+			continue
+		}
+		containerIPConfigs = append(containerIPConfigs, ip)
+	}
+
+	if len(containerIPConfigs) == 0 {
+		return fmt.Errorf("got no container IPs")
+	}
+
+	for _, cip := range containerIPConfigs {
+		containerNs, err := ns.GetNS(conf.PrevResult.Interfaces[*cip.Interface].Sandbox)
+		if err != nil {
+			return err
+		}
+		err = containerNs.Do(func(_ ns.NetNS) error {
+			return CreateIngressQdisc(conf.Rate, conf.Burst, conf.PrevResult.Interfaces[*cip.Interface].Name)
+		})
+		if err != nil {
+			return err
+		}
+
+		for pidx, prevIface := range conf.PrevResult.Interfaces {
+			if *cip.Interface == pidx {
+				continue
+			}
+			nonContainerNetNs, err := ns.GetNS(prevIface.Sandbox)
+			if err != nil {
+				return err
+			}
+			err = nonContainerNetNs.Do(func(netNS ns.NetNS) error {
+				nonContainerIfaceName := getRoutableHostIF(cip.Address.IP)
+				if nonContainerIfaceName != "" {
+					ifbDeviceName := os.Getenv("CNI_IFNAME")
+					mtu, err := getMTU(nonContainerIfaceName)
+					if err != nil {
+						return err
+					}
+					err = CreateIfb(ifbDeviceName, mtu)
+					if err != nil {
+						return err
+					}
+
+					ifbDevice, err := netlink.LinkByName(ifbDeviceName)
+					if err != nil {
+						return err
+					}
+
+					conf.PrevResult.Interfaces = append(conf.PrevResult.Interfaces, &current.Interface{
+						Name:    ifbDeviceName,
+						Sandbox: prevIface.Sandbox,
+						Mac:     ifbDevice.Attrs().HardwareAddr.String(),
+					})
+					return CreateEgressQdisc(conf.Rate, conf.Burst, nonContainerIfaceName, ifbDeviceName)
+				}
+
+				return nil
+			})
+
+			if err != nil {
+				return err
+			}
+		}
+	}
+
+	// Pass through the result for the next plugin
+	return types.PrintResult(conf.PrevResult, conf.CNIVersion)
+}
+func getMTU(deviceName string) (int, error) {
+	link, err := netlink.LinkByName(deviceName)
+	if err != nil {
+		return -1, err
+	}
+
+	return link.Attrs().MTU, nil
+}
+
+// cmdDel is called for DELETE requests
+func cmdDel(args *skel.CmdArgs) error {
+	conf, err := parseConfig(args.StdinData)
+	if err != nil {
+		return err
+	}
+	_ = conf
+
+	for _, prevIface := range conf.PrevResult.Interfaces {
```

Comment:
Ah, I'm afraid `PrevResult` will not be populated on a `DEL` call.  The [runtime will execute the plugins in reverse order](https://github.com/containernetworking/cni/blob/869505bebdf9/libcni/api.go#L147).  (also, to top it off, most plugins don't return any data on a successful `DEL`).

So to know the interface name to teardown, you may need some other tricks.  One way is that `ADD` may store some state (i.e. the interface name) on the filesystem, and `DEL` reads it back out.  See the other comment about the IFB device name.

Reply:
Dealt with this with generating a deterministic unique name (sha1 of network name and container id).